jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java

```diff
@@ -0,0 +1,266 @@
+    }
+
+    /**
+     * Getter for the next free Integer. Supplies the add vertex methods with new vertices
+     * 
+     * @return unused integer
+     */
+    private Integer getNextInteger()
+    {
+        return nextInteger++;
+    }
+
+    /**
+     * Method for adding a new join node.<br>
+     * {@code node} is copied two times jLeft and jRight. {@code node} afterwards becomes the root
+     * of the subtree. jRight becomes the child of {@code node} and has the children of {@code node}
+     * as children. jLeft becomes a leaf and the child of {@code node}. This can be used to make a
+     * join node retrospectively to branch of this node. jRight continues the path while jLeft adds
+     * another path.<br>
+     * The time complexity of this method is in $\mathcal{O}(|b(node)|)$.
+     * 
+     * @param node which nodes should become a join node
+     * @return the new children of the join node; first/left element has no children, second/right
+     *         element has the children of {@code node}
+     */
+    protected Pair<Integer, Integer> addJoin(Integer node)
+    {
+        Set<V> currentVertexBag = null;
+
+        // create first/left new child of node
+        Integer vertexChildLeft = getNextInteger();
+        decomposition.addVertex(vertexChildLeft);
+        currentVertexBag = new HashSet<V>(decompositionMap.get(node));
+        decompositionMap.put(vertexChildLeft, currentVertexBag);
+
+        // create second/right new child of node
+        Integer vertexChildRight = getNextInteger();
+        decomposition.addVertex(vertexChildRight);
+        currentVertexBag = new HashSet<V>(decompositionMap.get(node));
+        decompositionMap.put(vertexChildRight, currentVertexBag);
+
+        // redirect all edges to new parent, the second/right element
+        for (Integer successor : Graphs.successorListOf(decomposition, node)) {
+            decomposition.removeEdge(node, successor);
+            decomposition.addEdge(vertexChildRight, successor);
+        }
+        // make edge from node to left and right vertex
+        decomposition.addEdge(node, vertexChildLeft);
+        decomposition.addEdge(node, vertexChildRight);
+
+        return new Pair<Integer, Integer>(vertexChildLeft, vertexChildRight);
+    }
+
+    /**
+     * Method for adding forget nodes. It is only usable if {@code node} is a leaf. It then adds the
+     * new node as the child of {@code node} with the set of {@code node} plus
+     * {@code forgottenElement}.<br>
+     * The time complexity of this method is in $\mathcal{O}(|b(node)|)$.
+     * 
+     * @param forgottenElement the element, which gets forgotten
+     * @param node the node of the tree decomposition, which becomes a forget node
+     * @return the newly created node; or null and no change if either {@code forgottenElement} is
+     *         in b({@code node}) or {@code node} is not a leaf.
+     */
+    protected Integer addForget(V forgottenElement, Integer node)
```

Comment:
As far as I can tell you use the method only once in `ChordalNiceDecompositionBuilder`. There you have this line: `forgetNodeMap.put(vertex, decompNode);` If you return `null` then `vertex` will be removed from the map. Is that the intended behaviour? If this was by design than I guess you can use `Integer` here.